DigitalLearningSolutions.Web/Controllers/MyAccountController.cs

```diff
@@ -49,23 +56,83 @@ public IActionResult EditDetails()
+                "save" => EditDetailsPostSave(model),
+                "previewImage" => EditDetailsPostPreviewImage(model),
+                "removeImage" => EditDetailsPostRemoveImage(model),
+                _ => View(model),
+            };
+        }
+
+        private IActionResult EditDetailsPostSave(EditDetailsViewModel model)
         {
             if (!ModelState.IsValid)
             {
                 return View(model);
             }
 
+            if (model.ProfilePicture != null)
+            {
+                ModelState.AddModelError("ProfilePicture", "Preview your new profile picture before saving");
+                return View(model);
+            }
+
             var userAdminId = User.GetAdminId();
             var userDelegateId = User.GetNullableCandidateId();
 
-            if (!userService.TryUpdateUserAccountDetails(userAdminId, userDelegateId, model.Password,  model.FirstName, model.LastName, model.Email))
+            var profileImageToSave = model.HasProfileImageBeenRemoved ? null : model.NewProfileImage ?? model.CurrentProfileImage;
+
+            if (!userService.TryUpdateUserAccountDetails(userAdminId, userDelegateId, model.Password,  model.FirstName, model.LastName, model.Email, profileImageToSave))
             {
                 ModelState.AddModelError("Password", "The password you have entered is incorrect.");
                 return View(model);
             }
 
             return RedirectToAction("Index");
         }
+
+        private IActionResult EditDetailsPostPreviewImage(EditDetailsViewModel model)
+        {
+            // We don't want to display validation errors on other fields in this case
+            foreach (var key in ModelState.Keys.Where(k => k != "ProfilePicture"))
+            {
+                ModelState[key].Errors.Clear();
+                ModelState[key].ValidationState = ModelValidationState.Valid;
+            }
+
+            if (!ModelState.IsValid)
+            {
+                return View(model);
+            }
+
+            // Remove the old new profile image value from the ModelState
+            ModelState.Remove("NewProfileImage");
+            if (model.ProfilePicture != null)
+            {
+                model.NewProfileImage = imageResizeService.ResizeProfilePicture(model.ProfilePicture);
+            }
+
+            return View(model);
+        }
+
+        private IActionResult EditDetailsPostRemoveImage(EditDetailsViewModel model)
+        {
+            // We don't want to display validation errors on other fields in this case
+            foreach (var key in ModelState.Keys)
+            {
+                ModelState[key].Errors.Clear();
+                ModelState[key].ValidationState = ModelValidationState.Valid;
+            }
+            
+            ModelState.Remove("HasProfileImageBeenRemoved");
+            model.ProfilePicture = null;
+            model.NewProfileImage = null;
+            model.HasProfileImageBeenRemoved = true;

```

Comment:
Do we need both "HasProfileImageBeenRemoved" and "ProfileImage" fields? Wouldn't ProfileImage be null if it's removed or do we have a different case to this?

Reply:
Needed to distinguish between the 2 cases where NewProfileImage is null, when the user hasn't attempted to upload a new profile image before saving (they like their current one) and the case when the user has actively removed their profile picture.

All the logic here might be possible to refactor so I don't need the Current/New ProfileImages or this bool, but I'm not sure whether it will catch all cases easily

Reply:
Oh, do you mean like a "Reset to old picture" logic? I suspect we want the user to always "Preview" if they've selected a new file, and if they haven't, they can click "Remove" or upload and preview a different file.
If we want the reset old picture functionality in the future, we can add it as future work, but it doesn't need to be handled now.

Reply:
Nope, the NewProfileImage (now renamed) byte[]? property is null until the user goes through the preview action when they want to submit a new image.
It is also null if they do nothing relating to images (i.e. keep old image)
And the third case, when the user clears the profile picture, it is also null.

We need to distinguish between the different cases when this property is null image removed vs image never set. i.e. remove profile picture and do nothing.